Gunicorn webserver restarts every 30s

```
Feb 11 21:13:28 server3 python[2067604]: [2024-02-11 21:13:28 +0100] [2067604] [CRITICAL] WORKER TIMEOUT (pid:2098956)
Feb 11 21:13:28 server3 python[2098956]: [2024-02-11 21:13:28 +0100] [2098956] [INFO] Worker exiting (pid: 2098956)
Feb 11 21:13:29 server3 python[2067604]: [2024-02-11 21:13:29 +0100] [2067604] [ERROR] Worker (pid:2098956) was sent SIGKILL! Perhaps out of memory?
Feb 11 21:13:29 server3 python[2099723]: [2024-02-11 21:13:29 +0100] [2099723] [INFO] Booting worker with pid: 2099723
F
```

--- a/main.py
+++ b/main.py
@@ -5,7 +5,7 @@ import threading
 from src.services.watchdog_service import FolderMonitor
 from src.services.ocr_service import OcrService
 from src.services.sync_service import SyncService
-from src.services.flask_service import start_server, start_dev_server
+from src.services.flask_service import start_dev_server, start_socketio_server
 from src.webserver.dashboard import websocket_messages_queue
 from src.services.smb_server_alternative import SambaController
 from queue import Queue
@@ -54,16 +54,6 @@ if __name__ == "__main__":
     ocr_queue = Queue()
     sync_queue = Queue()
 
-    # Setup SMB
-
-    # This is for the smb_service_old
-    # smb_settings = config.get("smb_service")
-
-    # if args.smb_port:
-    #     smb_server = MySMBServer(smb_settings, args.smb_port)
-    # else:
-    #     smb_server = MySMBServer(smb_settings)
-
     # New SMB setup
     smb = SambaController()
     if not smb.check_share_exists(config.get("smb_service.share_name")):
@@ -78,18 +68,16 @@ if __name__ == "__main__":
     watchdog_thread = threading.Thread(target=watchdog.start_monitoring, name="Watchdog Service")
     ocr_thread = threading.Thread(target=ocr_service.start_processing, name="OCR Service")
     sync_thread = threading.Thread(target=sync_service.start_processing, name="Sync Service")
-    # smb_thread = threading.Thread(target=smb_server.start, name="SMB Service")
 
     if args.dev:
         logger.warning("Running FLASK dev server!")
         flask_thread = threading.Thread(target=start_dev_server, name="Flask Service DEV")
     else:
-        flask_thread = threading.Thread(target=start_server, name="Gunicorn Service")
+        flask_thread = threading.Thread(target=start_socketio_server, name="SocketIO Service")
 
     watchdog_thread.start()
     ocr_thread.start()
     sync_thread.start()
-    # smb_thread.start()
     flask_thread.start()
 
     try:
@@ -114,7 +102,5 @@ if __name__ == "__main__":
     flask_thread.join()
     logger.debug("Flask thread joined")
     smb.stop_server()
-    # smb_server.stop()
-    # smb_thread.is_alive()
-    # smb_thread.join()
+
     logger.info("PySyncOCR stopped. Bye!")

--- a/requirements.txt
+++ b/requirements.txt
@@ -12,7 +12,7 @@ dnspython==2.4.2
 dsinternals==1.2.4
 exceptiongroup==1.2.0
 Flask==3.0.0
-Flask-SocketIO=5.3.6
+Flask-SocketIO==5.3.6
 future==0.18.3
 gevent==23.9.1
 gevent-websocket == 0.10.1

--- a/requirements.txt
+++ b/requirements.txt
@@ -1,4 +1,5 @@
 beautifulsoup4==4.12.3
+bidict==0.22.1
 blinker==1.7.0
 cffi==1.16.0
 charset-normalizer==3.3.2
@@ -13,6 +14,8 @@ exceptiongroup==1.2.0
 Flask==3.0.0
 flask-sock==0.7.0
 future==0.18.3
+gevent==23.9.1
+gevent-websocket == 0.10.1
 gunicorn==21.2.0
 h11==0.14.0
 humanfriendly==10.0

--- a/requirements.txt
+++ b/requirements.txt
@@ -12,7 +12,7 @@ dnspython==2.4.2
 dsinternals==1.2.4
 exceptiongroup==1.2.0
 Flask==3.0.0
-flask-sock==0.7.0
+Flask-SocketIO=5.3.6
 future==0.18.3
 gevent==23.9.1
 gevent-websocket == 0.10.1

--- a/src/configs/config.json
+++ b/src/configs/config.json
@@ -1,6 +1,7 @@
 {
     "sync_service": {
         "root_folder": "/home/ocr/pysyncocr_shared",
+        "user_path": "/home/ocr",
         "failed_dir": "/home/ocr/pysyncocr_shared/failed",
         "keep_original": false,
         "original_dir": "/home/ocr/pysyncocr_shared/original"

--- a/src/helpers/config.py
+++ b/src/helpers/config.py
@@ -28,7 +28,11 @@ class Config:
             for k in keys:
                 value = value.get(k, {})
             # logger.debug(f"Read config key {key} with value {value} of type {type(value)}")
-            return value if value is not None else default
+            if value is not None:
+                return value
+            else:
+                logger.warning(f"Key {key} not found in config")
+                return default
         except Exception as ex:
             logger.exception(f"Failed reading config: {ex}")
             return default
@@ -39,7 +43,7 @@ class Config:
             value = self._config
             for k in keys:
                 value = value.get(k, {})
-            logger.debug(f"Read config key {key} with value {value} of type {type(value)}")
+            # logger.debug(f"Read config key {key} with value {value} of type {type(value)}")
             if value is not None:
                 return join(expanduser('~'), value)
             else:

--- a/src/helpers/rclone_configManager.py
+++ b/src/helpers/rclone_configManager.py
@@ -1,4 +1,5 @@
 import json
+import subprocess
 from src.helpers.logger import logger
 from src.helpers.config import config
 import os
@@ -73,10 +74,26 @@ class ConfigManager:
 
                 # Add new smb folder
                 try:
-                    os.makedirs(os.path.join(config.get_filepath("smb_service.share_path"), local), exist_ok=True)
-                    logger.info(f"Created folder {os.path.join(config.get_filepath('smb_service.share_path'),local)}")
+                    # Create folder
+                    folder_to_create = os.path.join(config.get_filepath("smb_service.share_path"), local)
+                    os.makedirs(folder_to_create, exist_ok=True, mode=775)
+                    logger.info(f"Created folder {folder_to_create}")
+
+                    # Set owner of folder to user "ocr"
+                    command = ["sudo", "chown", "-R", "ocr:ocr", folder_to_create]
+                    logger.debug(f"Calling {' '.join(command)}")
+                    res = subprocess.run(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, check=True)
+                    logger.debug(f"Received {res}")
+                    logger.info(f"Set owner of {folder_to_create} to user ocr.")
+
+                    # Set permissions of folder to user "ocr"
+                    command = ["sudo", "chmod", "-R", "775", folder_to_create]
+                    logger.debug(f"Calling {' '.join(command)}")
+                    res = subprocess.run(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, check=True)
+                    logger.debug(f"Received {res}")
+                    logger.info(f"Set permissions of {folder_to_create} to user ocr.")
                 except Exception as ex:
-                    logger.exception(f"Failed creating folder {os.path.join(config.get_filepath('smb_service.share_path'),local)}:", ex)
+                    logger.exception(f"Failed creating folder {folder_to_create}: {ex}")
                     return False
                 self.config_data.append(new_entry)
                 self.save_config()

--- a/src/helpers/rclone_management_onedrive.py
+++ b/src/helpers/rclone_management_onedrive.py
@@ -20,14 +20,16 @@ def list_remotes() -> list[str]:
         result = subprocess.run(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, check=True)
         logger.debug(f"Received {result}")
 
-        # Split the input string by lines and remove empty lines
-        lines = [line.strip() for line in result.stdout.decode().split('\n') if line.strip()]
+        # Split the data by newline characters to get individual lines
+        lines = result.stdout.decode().split('\n')
 
-        # Remove colons from each line
-        filtered_list = [item for item in lines if ": onedrive" in item]
-        result_list = [line.replace(': onedrive', '') for line in filtered_list]
-        logger.debug(f"Found remotes: {str(result_list)}")
-        return result_list
+        # Remove additional spaces from the values
+        result2 = [value.replace(' ', '') for value in lines]
+
+        # Use list comprehension to filter lines and extract values before ':'
+        result = [line.split(':')[0].strip() for line in result2 if ':onedrive' in line]
+        logger.debug(f"Found remotes: {str(result)}")
+        return result
     except subprocess.CalledProcessError as e:
         logger.error(f"Error getting remotes rclone: {e}")
         logger.error(e.stderr.decode())

--- a/src/helpers/rclone_setup.py
+++ b/src/helpers/rclone_setup.py
@@ -3,6 +3,7 @@ import re
 from src.helpers.logger import logger
 from src.helpers.run_subprocess import run_subprocess
 from flask import current_app
+from src.webserver import socketio
 
 
 def extract_url(prompt):
@@ -17,6 +18,8 @@ def configure_rclone_onedrive_personal(name):
     # Run the rclone config command using pexpect
     process = pexpect.spawn('rclone config')
     logger.info("Configuring OneDrive with rclone, please wait...")
+    socketio.emit("message_update", "Connecting to OneDrive...", namespace="/websocket-onedrive")
+    socketio.sleep(0.1)
     try:
         # Wait for the initial output and print it
         process.expect("Quit config", timeout=10)
@@ -37,10 +40,12 @@ def configure_rclone_onedrive_personal(name):
         except pexpect.exceptions.TIMEOUT:
             if "already exists" in process.buffer.decode():
                 logger.error("Remote with that name already exists.")
-                return -2
+                socketio.emit("message_update", "Remote with that name already exists.", namespace="/websocket-onedrive")
+                socketio.sleep(0.1)
             else:
                 logger.error(process.buffer.decode())
-                return -1
+                socketio.emit("message_update", "Error: " + process.buffer.decode(), namespace="/websocket-onedrive")
+                socketio.sleep(0.1)
         process.sendline('onedrive')
         logger.debug("Selected option onedrive")
 
@@ -71,7 +76,8 @@ def configure_rclone_onedrive_personal(name):
         if auth_url:
             logger.info("If your browser doesn't open automatically, go to the following link:")
             logger.info(auth_url)
-            yield auth_url
+            socketio.emit("message_update", auth_url, namespace="/websocket-onedrive")
+            socketio.sleep(0.1)
 
             # Continue handling prompts
             process.expect('Your choice>', timeout=120)
@@ -93,19 +99,23 @@ def configure_rclone_onedrive_personal(name):
             process.sendline('y')
             logger.debug("Selected [y] for final confirmation")
             logger.info("Success setting up OneDrive!")
-            yield "0"
+            socketio.emit("message_update", "Success setting up OneDrive!", namespace="/websocket-onedrive")
+            socketio.sleep(0.1)
         else:
             logger.error("Failed to extract authorization URL.")
-            return -3
+            socketio.emit("message_update", "Failed to extract authorization URL.", namespace="/websocket-onedrive")
+            socketio.sleep(0.1)
 
     except pexpect.EOF:
         logger.error("Error: Unexpected end of input.")
         logger.error(process.before.decode())
-        return -4
+        socketio.emit("message_update", "Error: Unexpected end of input.", namespace="/websocket-onedrive")
+        socketio.sleep(0.1)
 
     except Exception:
         logger.error(process.buffer.decode())
-        return -5
+        socketio.emit("message_update", process.buffer.decode(), namespace="/websocket-onedrive")
+        socketio.sleep(0.1)
 
     finally:
         # Close the process

--- a/src/services/flask_service.py
+++ b/src/services/flask_service.py
@@ -17,7 +17,7 @@ def stop_server():
 def start_socketio_server():
     logger.info("Starting socketio server...")
     app = create_app()
-    socketio.run(app, debug=False, allow_unsafe_werkzeug=False, log_output=True)
+    socketio.run(app, debug=False, allow_unsafe_werkzeug=False, log_output=True, use_reloader=False)
 
 
 def start_dev_server():

--- a/src/services/flask_service.py
+++ b/src/services/flask_service.py
@@ -1,7 +1,6 @@
 from src.webserver import create_app
 from src.helpers.logger import logger
-import subprocess
-import os
+from src.webserver import socketio
 
 gunicorn_process = None  # Global variable to hold the gunicorn process
 stop_server_flag = False  # Flag to indicate when to stop the server
@@ -15,26 +14,10 @@ def stop_server():
         stop_server_flag = True
 
 
-def start_server():
-    global gunicorn_process, stop_server_flag
-    python_interpreter = os.path.join(os.getcwd(), '.venv', 'bin', 'python')
-
-    gunicorn_command = [
-        python_interpreter,
-        '-m', 'gunicorn',
-        '-w', '2',  # Number of worker processes
-        '-b', '0.0.0.0:5000',  # Host and port on which to bind
-        'src.webserver:create_app()',  # Replace with your actual app module and instance
-    ]
-
-    try:
-        gunicorn_process = subprocess.Popen(gunicorn_command)
-        gunicorn_process.wait()  # Wait for the process to finish
-    except Exception as e:
-        logger.error(f"Error running Gunicorn: {e}")
-    finally:
-        # Perform cleanup before exiting
-        logger.info("Gunicorn server stopped.")
+def start_socketio_server():
+    logger.info("Starting socketio server...")
+    app = create_app()
+    socketio.run(app, debug=False, allow_unsafe_werkzeug=False, log_output=True)
 
 
 def start_dev_server():

--- a/src/services/smb_server_alternative.py
+++ b/src/services/smb_server_alternative.py
@@ -49,7 +49,7 @@ class SambaController:
         try:
             if not os.path.exists(config.get_filepath("smb_service.share_path")):
                 os.mkdir(config.get_filepath("smb_service.share_path"), mode=777)
-                logger.info(f"Created directory {config.get_filepath('smb_server.share_path')}")
+                logger.info(f"Created directory {config.get_filepath('smb_service.share_path')}")
 
                 # Set owner of folder to user "ocr"
                 command = ["sudo", "chown", "-R", "ocr:ocr", config.get_filepath("smb_service.share_path")]
@@ -59,7 +59,7 @@ class SambaController:
                 logger.info(f"Set owner of {config.get_filepath('smb_service.share_path')} to user ocr.")
 
                 # Set permissions of folder to user "ocr"
-                command = ["sudo", "chmod", "-R", "755", config.get_filepath("smb_service.share_path")]
+                command = ["sudo", "chmod", "-R", "775", config.get_filepath("smb_service.share_path")]
                 logger.debug(f"Calling {' '.join(command)}")
                 res = subprocess.run(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, check=True)
                 logger.debug(f"Received {res}")
@@ -100,6 +100,27 @@ class SambaController:
                 logger.exception(f"Error adding user to Samba: {e}")
                 return
 
+        username = config.get("smb_service.username")
+        try:
+            if not os.path.exists(config.get("sync_service.user_path")):
+                os.mkdir(config.get("sync_service.user_path"), mode=755)
+                logger.info(f"Created directory {config.get('sync_service.user_path')}")
+                # Set owner of folder to user "ocr"
+                command = ["sudo", "chown", "-R", username + ":" + username, config.get("sync_service.user_path")]
+                logger.debug(f"Calling {' '.join(command)}")
+                res = subprocess.run(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, check=True)
+                logger.debug(f"Received {res}")
+                logger.info(f"Set owner of {config.get('sync_service.user_path')} to user {username}.")
+                # Set permissions of folder to user "ocr"
+                command = ["sudo", "chmod", "-R", "755", config.get("sync_service.user_path")]
+                logger.debug(f"Calling {' '.join(command)}")
+                res = subprocess.run(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, check=True)
+                logger.debug(f"Received {res}")
+                logger.info(f"Set permissions of {config.get('sync_service.user_path')} to user {username}.")
+        except Exception as e:
+            logger.exception(f"Error creating home directory: {e}")
+            return
+
         # Run sudo smbpasswd -a username
         try:
             process = subprocess.Popen(["sudo", "smbpasswd", "-a", username],
@@ -136,7 +157,7 @@ class SambaController:
                     if line.strip().startswith(f"[{share_name}]"):
                         logger.debug(f"Share {share_name} exists.")
                         return True
-            logger.debug(f"Share {share_name} does not exist.")
+            logger.info(f"Share {share_name} does not exist.")
             return False
         except Exception as e:
             logger.exception(f"Error checking Samba share: {e}")

--- a/src/services/watchdog_service.py
+++ b/src/services/watchdog_service.py
@@ -252,7 +252,7 @@ class FolderMonitor:
         for root, dirs, files in os.walk(source_dir):
             # Remove the "failed" and "done" directories from the search
             if "failed" in dirs:
-                dirs.remove("failed") 
+                dirs.remove("failed")
             if "done" in dirs:
                 dirs.remove("done")
 

--- a/src/services/watchdog_service.py
+++ b/src/services/watchdog_service.py
@@ -13,7 +13,7 @@ from PyPDF2 import PdfReader
 from src.helpers.ProcessItem import ItemType, ProcessItem, ProcessStatus
 from src.helpers.rclone_configManager import RcloneConfig
 import fitz
-from src.webserver.db import update_scanneddata_database
+from src.webserver.db import send_database_request, update_scanneddata_database
 
 
 class FileHandler(FileSystemEventHandler):
@@ -215,7 +215,7 @@ class FolderMonitor:
             self.observer.schedule(self.event_handler, self.root_folder, recursive=True)
             self.observer.start()
         except Exception as e:
-            logger.exception("Failed starting watchdog", e)
+            logger.exception(f"Failed starting watchdog: {e}")
             return
 
         # Check for errors
@@ -252,7 +252,7 @@ class FolderMonitor:
         for root, dirs, files in os.walk(source_dir):
             # Remove the "failed" and "done" directories from the search
             if "failed" in dirs:
-                dirs.remove("failed")
+                dirs.remove("failed") 
             if "done" in dirs:
                 dirs.remove("done")
 
@@ -263,9 +263,17 @@ class FolderMonitor:
                     if file.lower().endswith(".pdf"):
                         try:
                             source_file = os.path.join(root, file)
+                            source_file_top_dir = os.path.basename(os.path.dirname(source_file))
+                            source_file_filename = os.path.basename(source_file)
                             destination_file = os.path.join(failed_dir, file)
                             shutil.move(source_file, destination_file)
                             logger.warning(f"Moved '{file}' to '{failed_dir}'")
+
+                            # Now try to update the db item to failed
+                            try:
+                                send_database_request('UPDATE ' + config.get("sql.db_pdf_table") + ' SET file_status="' + str(ProcessStatus.FAILED.value) + '" WHERE local_filepath="' + source_file_top_dir + '" AND file_name="' + source_file_filename + '"')
+                            except Exception as ex:
+                                logger.exception(f"Failed updating database item to failed: {ex}")
                         except Exception as ex:
                             logger.exception(f"Failed moving '{file}' to '{failed_dir}': {ex}")
 

--- a/src/webserver/__init__.py
+++ b/src/webserver/__init__.py
@@ -1,11 +1,11 @@
 import os
-from flask_sock import Sock
 from flask import Flask
 from src.helpers.logger import logger
 from src.webserver.context_processor import inject_template_data
 from . import root
+from flask_socketio import SocketIO
 
-sock = Sock()
+socketio = SocketIO(async_mode='threading')
 
 
 def create_app(test_config=None):
@@ -30,7 +30,7 @@ def create_app(test_config=None):
         pass
 
     # initialize websocket
-    sock.init_app(app)
+    socketio.init_app(app)
 
     # initialize SQL db
     from . import db

--- a/src/webserver/__init__.py
+++ b/src/webserver/__init__.py
@@ -5,7 +5,7 @@ from src.webserver.context_processor import inject_template_data
 from . import root
 from flask_socketio import SocketIO
 
-socketio = SocketIO(async_mode='threading')
+socketio = SocketIO()
 
 
 def create_app(test_config=None):

--- a/src/webserver/dashboard.py
+++ b/src/webserver/dashboard.py
@@ -2,14 +2,14 @@ import json
 from flask import Blueprint, render_template, request, current_app
 bp = Blueprint('dashboard', __name__, url_prefix='/dashboard')
 from src.helpers.logger import logger
-from src.webserver.db import get_db
+from src.webserver.db import get_db, send_database_request, send_database_request_with_headers
 from src.helpers.config import config
 from src.helpers.time_converter import format_time_difference
 import locale
 from datetime import datetime
 import time
 import math
-from . import sock
+from . import socketio
 import queue
 
 websocket_messages_queue = queue.Queue()
@@ -143,8 +143,19 @@ def index():
                                latest_timestamp_completed_string="Unknown")
 
 
-@sock.route("/websocket")
-def websocket_dashboard(ws):
+@socketio.on('connect', namespace='/websocket')
+def websocket_connect():
+    logger.debug("Client connected to websocket")
+    socketio.start_background_task(target=websocket_dashboard_loop)
+    # emit('message_update', json.dumps({'data': 'Connected', 'count': 0}))
+
+
+@socketio.on('disconnect', namespace='/websocket')
+def websocket_disconnect():
+    logger.debug("Client disconnected from websocket")
+
+
+def websocket_dashboard_loop():
     while True:
         try:
             # Check if there's a message in the queue
@@ -157,13 +168,12 @@ def websocket_dashboard(ws):
                 continue
 
             db_id = signal['id']
-            db = get_db()
-            pdf_list = db.execute(
+            pdf_list = send_database_request_with_headers(
                 'SELECT *, DATETIME(created, "localtime") AS local_created, DATETIME(modified, "localtime") AS local_modified FROM scanneddata '
-                'WHERE id = :id',
-                {'id': db_id}
-            ).fetchall()
-            pdf = [dict(pdf) for pdf in pdf_list][0]
+                'WHERE id = {}'.format(db_id)
+            )
+
+            pdf = pdf_list[0]
             logger.debug(f"Loaded {db_id} pdf from db")
             try:
                 input_datetime_created = datetime.strptime(pdf['local_created'], "%Y-%m-%d %H:%M:%S")
@@ -174,24 +184,24 @@ def websocket_dashboard(ws):
                 logger.exception(f"Failed setting datetime for {pdf['id']}. {ex}")
             data_to_send = {'id': pdf['id'], 'command': signal['command'], 'data': pdf}
             logger.debug(f"Sending update data {data_to_send} to websocket")
-            ws.send(json.dumps(data_to_send))
+            socketio.emit('message_update', json.dumps(data_to_send), namespace="/websocket")
 
             # Now check the number of pending and finished docs to update dashboard
             try:
-                pending_pdfs = db.execute(
+                pending_pdfs = send_database_request(
                     'SELECT COUNT(*) FROM scanneddata '
                     'WHERE LOWER(file_status) LIKE "%pending%"'
-                    ).fetchone()[0]
+                    )[0][0]
                 logger.debug(f"Found {pending_pdfs} queued pdfs")
             except Exception as e:
                 logger.exception(f"Error getting pending pdf count: {e}")
                 pending_pdfs = "Unknown"
 
             try:
-                processed_pdfs = db.execute(
+                processed_pdfs = send_database_request(
                             'SELECT COUNT(*) FROM scanneddata '
                             'WHERE file_status = "Completed" '
-                            ).fetchone()[0]
+                            )[0][0]
                 logger.debug(f"Found {processed_pdfs} completed pdfs")
             except Exception as e:
                 logger.exception(f"Error getting processed pdf count: {e}")
@@ -200,12 +210,12 @@ def websocket_dashboard(ws):
             # Now get the latest timestamps for the dashboard
             # Get the latest timestamp from the file_status=pending
             try:
-                latest_timestamp_pending = db.execute(
+                latest_timestamp_pending = send_database_request(
                     'SELECT DATETIME(created, "localtime") FROM scanneddata '
                     'WHERE file_status = "Pending" '
                     'ORDER BY created DESC '
                     'LIMIT 1'
-                    ).fetchone()
+                    )[0]
                 if latest_timestamp_pending is not None:
                     logger.debug(f"Found latest timestamp for pending documents: {latest_timestamp_pending[0]}")
                     latest_timestamp_pending_string = "Updated " + format_time_difference(latest_timestamp_pending[0])
@@ -218,12 +228,12 @@ def websocket_dashboard(ws):
 
             # Get the latest timestamp from the file_status=completed
             try:
-                latest_timestamp_completed = db.execute(
+                latest_timestamp_completed = send_database_request(
                     'SELECT DATETIME(modified, "localtime") FROM scanneddata '
                     'WHERE file_status = "Completed" '
                     'ORDER BY modified DESC '
                     'LIMIT 1'
-                    ).fetchone()
+                    )[0]
                 if latest_timestamp_completed is not None:
                     logger.debug(f"Found latest timestamp for synced documents: {latest_timestamp_completed[0]}")
                     latest_timestamp_completed_string = "Updated " + format_time_difference(latest_timestamp_completed[0])
@@ -239,7 +249,7 @@ def websocket_dashboard(ws):
                                                                     'pending_pdfs_latest_timestamp': latest_timestamp_pending_string,
                                                                     'processed_pdfs_latest_timestamp': latest_timestamp_completed_string}}
             logger.debug(f"Sending update dashboard data {data_to_send} to websocket")
-            ws.send(json.dumps(data_to_send))
+            socketio.emit('message_update', json.dumps(data_to_send), namespace="/websocket")
             logger.debug("Sent update dashboard command to websocket")
             logger.debug("Waiting for next signal")
         except queue.Empty:

--- a/src/webserver/dashboard.py
+++ b/src/webserver/dashboard.py
@@ -7,10 +7,10 @@ from src.helpers.config import config
 from src.helpers.time_converter import format_time_difference
 import locale
 from datetime import datetime
-import time
 import math
 from . import socketio
 import queue
+
 
 websocket_messages_queue = queue.Queue()
 
@@ -147,7 +147,6 @@ def index():
 def websocket_connect():
     logger.debug("Client connected to websocket")
     socketio.start_background_task(target=websocket_dashboard_loop)
-    # emit('message_update', json.dumps({'data': 'Connected', 'count': 0}))
 
 
 @socketio.on('disconnect', namespace='/websocket')
@@ -156,6 +155,8 @@ def websocket_disconnect():
 
 
 def websocket_dashboard_loop():
+    logger.info("Started websocket dashboard loop")
+
     while True:
         try:
             # Check if there's a message in the queue
@@ -254,4 +255,4 @@ def websocket_dashboard_loop():
             logger.debug("Waiting for next signal")
         except queue.Empty:
             pass
-        time.sleep(0.2)
+        socketio.sleep(0.2)

--- a/src/webserver/db.py
+++ b/src/webserver/db.py
@@ -114,3 +114,47 @@ def send_database_request(query: str):
         return result
     except Exception as ex:
         logger.exception(f"Error sending database request: {ex}")
+
+
+def send_database_request_with_headers(query: str) -> list[dict]:
+    """Sends a request to the database and returns the result.
+
+    Connects to the SQLite database, executes the provided query,
+    and returns the result.
+    Commits the changes and closes the connection.
+    Returns list of dict with row descriptions.
+
+    Handles any errors and logs exceptions.
+    """
+    try:
+        # Connect to the database
+        connection = sqlite3.connect(config.get("sql.db_location"))
+
+        # Create a cursor object
+        cursor = connection.cursor()
+
+        # Execute the query
+        logger.debug("Sending database request: " + query)
+        cursor.execute(query)
+
+        # Fetch all the results
+        rows = cursor.fetchall()
+
+        # Fetch the column names from the cursor description
+        columns = [col[0] for col in cursor.description]
+
+        # Convert the rows into dictionaries
+        result_dicts = []
+        for row in rows:
+            result_dicts.append(dict(zip(columns, row)))
+
+        # Commit the changes
+        connection.commit()
+
+        # Close the connection
+        connection.close()
+
+        return result_dicts
+    except Exception as ex:
+        logger.exception(f"Error sending database request: {ex}")
+        return None

--- a/src/webserver/static/js/dashboard.js
+++ b/src/webserver/static/js/dashboard.js
@@ -5,12 +5,11 @@ document.addEventListener('DOMContentLoaded', function () {
         loadinglottieanimation.style.display = 'none';
     });
 
-    const socket = new WebSocket('ws://' + window.location.host + '/websocket');
-    socket.addEventListener('message', ev => {
-        console.log("Received" + ev.data);
-
+    var socket = io.connect('http://' + window.location.host + '/websocket');
+    socket.on('message_update', function(data) {
         // convert ev.data to josn object
-        var jsonData = JSON.parse(ev.data);
+        var jsonData = JSON.parse(data);
+        console.log("Received message: " + data);
 
         // Test if command property if jsonData is equal to 'update'
         if (jsonData.command === 'update') {
@@ -26,10 +25,12 @@ document.addEventListener('DOMContentLoaded', function () {
             console.log("Received update_dashboard command");
             // If it is, update the dashboard
             updateDashboard(jsonData.data);
+        } else {
+            console.log("Unknown command received.");
         }
 
     });
-    socket.addEventListener('open', ev => {
+    socket.on('connect', function () {
         console.log('Connected to websocket');
     });
 

--- a/src/webserver/static/js/sync.js
+++ b/src/webserver/static/js/sync.js
@@ -61,9 +61,7 @@ function evaluatePathMappingSubmitButton() {
     const remotePathInput = document.getElementById('remote_path');
 
     // Enable or disable the button based on whether the input has content
-    submitButtonpathmapping.disabled = !localpathinput.value.trim();
-    submitButtonpathmapping.disabled = localpathinput.classList.contains('is-invalid');
-    submitButtonpathmapping.disabled = !remotePathInput.value.startsWith('/');
+    submitButtonpathmapping.disabled = !(localpathinput.value.trim() && !localpathinput.classList.contains('is-invalid') && remotePathInput.value.startsWith('/'));
 }
 
 function deleteOneDriveConf(id) {
@@ -131,26 +129,26 @@ function addOneDrive() {
                     sshTunnelInfoElement.style.display = "none;";
                     // Run websocket for status updates during rclone config
                     animation_statustext.innerText = "Waiting for connection...";
-                    const socket = new WebSocket('ws://' + window.location.host + '/websocket-onedrive');
-                    socket.addEventListener('message', ev => {
+                    var socket = io.connect('http://' + window.location.host + '/websocket-onedrive');
+                    socket.on('message_update', function (data) {
                         statusUpdateElement.style.display = "block";
-                        console.log(ev.data);
-                        // Test if ev.data begins with 'http'
-                        if (ev.data.startsWith("http")) {
+                        console.log("Received message: " + data);
+                        // Test if data begins with 'http'
+                        if (data.startsWith("http")) {
                             // If it does, it's a link to the file
-                            document.getElementById("updateText").innerHTML = '<a href="' + ev.data + '" target="_blank">' + "To authenticate, please visit <br>" + ev.data + '</a>';
-                            window.open(ev.data, '_blank');
-                        } else if (ev.data.startsWith("Success")) {
+                            document.getElementById("updateText").innerHTML = '<a href="' + data + '" target="_blank">' + "To authenticate, please visit <br>" + data + '</a>';
+                            window.open(data, '_blank');
+                        } else if (data.startsWith("Success")) {
                             window.location.reload();
                         } else {
                             // Otherwise, it's just text
                             animation.style.display = "none";
-                            document.getElementById("updateText").innerHTML = ev.data;
+                            document.getElementById("updateText").innerHTML = data;
                         }
                     });
-                    socket.addEventListener('open', ev => {
-                        console.log('Connected to websocket');
-                        socket.send(JSON.stringify({ "name": document.getElementById("onedrive_name").value }));
+                    socket.on('connect', function () {
+                        console.log('Connected to onedrive websocket');
+                        socket.emit('message_update', JSON.stringify({ "name": document.getElementById("onedrive_name").value }));
                     });
                 };
             };
@@ -436,6 +434,11 @@ function addPathMappingToUI(local, remote, connection) {
                 </div>
             </div>`
     document.getElementById("pathmappingscontainer").appendChild(newCard);
+
+    try {
+        const emptymsg = document.getElementById("emptypathmappingsmessage")
+        emptymsg.style.display = "none";
+    } catch {}
 }
 
 function deletePathMapping(id) {

--- a/src/webserver/static/js/sync.js
+++ b/src/webserver/static/js/sync.js
@@ -96,7 +96,6 @@ function addOneDrive() {
     const animation_statustext = document.getElementById("waitingAnimationonedriveadd_statustext");
     const form = document.getElementById("onedrive_name_container");
     const addButton = document.getElementById("add_onedrive_button");
-    const statusUpdateElement = document.getElementById("statusUpdate");
     const sshTunnelInfoElement = document.getElementById("sshTunnelSetupInfo");
     const sshInstallInfoElement = document.getElementById("sshInstallInfo");
     const cloud_header = document.getElementById("cloud_header");
@@ -125,25 +124,37 @@ function addOneDrive() {
                 // Show the ssh tunnel setup modal and reroute the next button
                 sshTunnelInfoElement.style.display = "block";
                 addButton.onclick = function () {
-                    animation.style.display = "block;";
-                    sshTunnelInfoElement.style.display = "none;";
+                    animation.style.display = "block";
+                    sshTunnelInfoElement.style.display = "none";
                     // Run websocket for status updates during rclone config
                     animation_statustext.innerText = "Waiting for connection...";
-                    var socket = io.connect('http://' + window.location.host + '/websocket-onedrive');
+                    addButton.style.display = "none";
+                    var rclonePopup = null;
+                    var socket = io.connect('http://' + window.location.host + '/websocket-onedrive', {reconnection: false});
                     socket.on('message_update', function (data) {
-                        statusUpdateElement.style.display = "block";
                         console.log("Received message: " + data);
                         // Test if data begins with 'http'
                         if (data.startsWith("http")) {
                             // If it does, it's a link to the file
-                            document.getElementById("updateText").innerHTML = '<a href="' + data + '" target="_blank">' + "To authenticate, please visit <br>" + data + '</a>';
-                            window.open(data, '_blank');
+                            animation_statustext.innerHTML = '<a href="#" onclick="openAuthWindow(\'' + data + '\')">To authenticate, please visit <br>' + data + '</a>';
+                            // openAuthWindow(data);
                         } else if (data.startsWith("Success")) {
+                            try {
+                                rCloneAuthPopup.close();
+                            } catch {}
                             window.location.reload();
                         } else {
                             // Otherwise, it's just text
-                            animation.style.display = "none";
-                            document.getElementById("updateText").innerHTML = data;
+                            if (data.startsWith("Error")) {
+                                const error_display = document.getElementById("errormsg_display");
+                                const errormsg = document.getElementById("error_msg_fromdisplay");
+                                errormsg.innerText = data;
+                                error_display.style.display = "block";
+                                animation.style.display = "none";
+                                addButton.innerText = "Retry";
+                            } else {
+                                animation_statustext.innerText = data;
+                            }
                         }
                     });
                     socket.on('connect', function () {
@@ -157,6 +168,9 @@ function addOneDrive() {
     xhr.send();
 }
 
+function openAuthWindow(url) {
+    rCloneAuthPopup = window.open(url, "rClone Authentication", "width=400,height=600");
+}
 
 function validateTextInput(input) {
     // Regular expression pattern for allowed characters

--- a/src/webserver/sync.py
+++ b/src/webserver/sync.py
@@ -5,7 +5,7 @@ from src.helpers.logger import logger
 from src.helpers.rclone_management_onedrive import dump_config, delete_config_item, list_remotes, list_folders
 from src.helpers.rclone_setup import configure_rclone_onedrive_personal, check_ssh_enabled
 from src.helpers.rclone_configManager import RcloneConfig
-from . import sock
+from . import socketio
 from src.webserver.db import get_db
 from src.helpers.config import config
 import json
@@ -149,35 +149,43 @@ def getOneDriveDirectories():
         return "Failed retrieving remotes", 500
 
 
-@sock.route("/websocket-onedrive")
-def websocket_onedrive(ws):
+@socketio.on('connect', namespace='/websocket-onedrive')
+def websocket_connect():
+    logger.debug("Client connected to onedrive websocket")
+
+
+@socketio.on('disconnect', namespace='/websocket-onedrive')
+def websocket_disconnect():
+    logger.debug("Client disconnected from onedrive websocket")
+
+
+@socketio.on('message_update', namespace='/websocket-onedrive')
+def handle_message(message):
     """
     Handle realtime sync updates from rclone sent over a websocket.
 
     Accepts a websocket connection and json data containing the onedrive remote name.
     Calls configure_rclone_onedrive_personal to sync the remote, and sends back sync updates over the websocket.
     """
-    while True:
-        data = ws.receive()
-        if data is None:
-            break
+    if message is None:
+        return
+    else:
+        logger.info(f"Received data: {message}")
+        json_data = json.loads(message)
+        if "name" in json_data:
+            try:
+                for update in configure_rclone_onedrive_personal(json_data["name"]):
+                    logger.debug(f"Received signal from rclone: {update}")
+                    if update.startswith("http"):
+                        socketio.emit('message_update', update, namespace='/websocket-onedrive')
+                    elif update == "0":
+                        socketio.emit('message_update', "Success: " + update, namespace='/websocket-onedrive')
+                    else:
+                        socketio.emit('message_update', "Failed: " + update, namespace='/websocket-onedrive')
+            except StopIteration as e:
+                socketio.emit('message_update', "An error occured: " + str(e.value) + "\n Please try again later.", namespace='/websocket-onedrive')
         else:
-            logger.info(f"Received data: {data}")
-            json_data = json.loads(data)
-            if "name" in json_data:
-                try:
-                    for update in configure_rclone_onedrive_personal(json_data["name"]):
-                        logger.debug(f"Received signal from rclone: {update}")
-                        if update.startswith("http"):
-                            ws.send(update)
-                        elif update == "0":
-                            ws.send("Success: " + update)
-                        else:
-                            ws.send("Failed: " + update)
-                except StopIteration as e:
-                    ws.send("An error occured: " + str(e.value) + "\n Please try again later.")
-            else:
-                ws.send("Failed - cant handle this.")
+            socketio.emit('message_update', "Failed - cant handle this.", message_namespace='/websocket-onedrive')
 
 
 @bp.post("/pathmapping")
@@ -195,6 +203,10 @@ def addPathMapping():
         json_data = request.get_json()
         if not json_data or "local_path" not in json_data or "remote_path" not in json_data or "remote_id" not in json_data:
             return "Missing required data in request", 400
+
+        if not json_data["local_path"] or not json_data["remote_path"] or not json_data["remote_id"]:
+            return "Values cannot be empty", 400
+
         logger.info(f"Received path mapping: {json_data}")
         res = RcloneConfig.add(json_data["local_path"], json_data["remote_path"], json_data["remote_id"])
         if res:

--- a/src/webserver/sync.py
+++ b/src/webserver/sync.py
@@ -152,6 +152,7 @@ def getOneDriveDirectories():
 @socketio.on('connect', namespace='/websocket-onedrive')
 def websocket_connect():
     logger.debug("Client connected to onedrive websocket")
+    socketio.emit('message_update', 'Connected to onedrive websocket', namespace='/websocket-onedrive')
 
 
 @socketio.on('disconnect', namespace='/websocket-onedrive')
@@ -172,20 +173,9 @@ def handle_message(message):
     else:
         logger.info(f"Received data: {message}")
         json_data = json.loads(message)
+        socketio.emit('message_update', 'Connecting to OneDrive...', namespace='/websocket-onedrive')
         if "name" in json_data:
-            try:
-                for update in configure_rclone_onedrive_personal(json_data["name"]):
-                    logger.debug(f"Received signal from rclone: {update}")
-                    if update.startswith("http"):
-                        socketio.emit('message_update', update, namespace='/websocket-onedrive')
-                    elif update == "0":
-                        socketio.emit('message_update', "Success: " + update, namespace='/websocket-onedrive')
-                    else:
-                        socketio.emit('message_update', "Failed: " + update, namespace='/websocket-onedrive')
-            except StopIteration as e:
-                socketio.emit('message_update', "An error occured: " + str(e.value) + "\n Please try again later.", namespace='/websocket-onedrive')
-        else:
-            socketio.emit('message_update', "Failed - cant handle this.", message_namespace='/websocket-onedrive')
+            socketio.start_background_task(target=configure_rclone_onedrive_personal, name=json_data["name"])
 
 
 @bp.post("/pathmapping")

--- a/src/webserver/templates/base.html
+++ b/src/webserver/templates/base.html
@@ -113,6 +113,7 @@
     <script src="https://code.jquery.com/jquery-3.7.1.slim.min.js"
         integrity="sha256-kmHvs0B+OpCW5GVHUNjv9rOmY0IvSIRcf7zGUDTDQM8=" crossorigin="anonymous"></script>
     <script src="https://unpkg.com/@dotlottie/player-component@latest/dist/dotlottie-player.mjs" type="module"></script>
+    <script src="https://cdn.socket.io/4.7.4/socket.io.min.js" integrity="sha384-Gr6Lu2Ajx28mzwyVR8CFkULdCU7kMlZ9UthllibdOSo6qAiN+yXNHqtgdTvFXMT4" crossorigin="anonymous"></script>
 </body>
 
 </html>

--- a/src/webserver/templates/sync.html
+++ b/src/webserver/templates/sync.html
@@ -272,9 +272,19 @@
                         </div>
                     </div>
 
-                    <!-- Status update -->
-                    <div class="text-center" id="statusUpdate" style="display: none;">
-                        <p class="mt-2" id="updateText">I am hidden</p>
+                    <!-- Error display -->
+                    <div id="errormsg_display" style="display: none;">
+                        <div id="lottieWarning" class="d-flex justify-content-center align-items-center">
+                            <dotlottie-player id="terminal_lottie_player" src="/static/lottie/error.json"
+                                background="transparent" speed="1" direction="1" playMode="normal" style="width: 150px; height: 150px;"
+                                autoplay></dotlottie-player>
+                        </div>
+                        <div class="text-center">
+                            <h5>Error</h5>
+                        </div>
+                        <p class="text-center">
+                            <span id="error_msg_fromdisplay"></span><br>
+                        </p>
                     </div>
                 </div>
                 <div class="modal-footer">

--- a/src/webserver/templates/sync.html
+++ b/src/webserver/templates/sync.html
@@ -92,7 +92,7 @@
         </div>
         {% endfor %}
         {% else %}
-        <p>No path mappings found.</p>
+        <p id="emptypathmappingsmessage">No path mappings found.</p>
         {% endif %}
     </div>
     <br>

--- a/tests/config/test_config.py
+++ b/tests/config/test_config.py
@@ -48,7 +48,7 @@ def test_iter(conf_setup):
 
 
 def test_get(conf_setup):
-    assert conf_setup.get("sync_service.root_folder") == "PySyncOCR"
+    assert conf_setup.get("sync_service.root_folder") == "/home/ocr/pysyncocr_shared"
 
 
 def test_get_empty(conf_setup):
@@ -67,9 +67,3 @@ def test_set(conf_setup):
 def test_attr(conf_setup):
     assert type(conf_setup.get("sync_service.root_folder")) == str
     assert type(conf_setup.get("sync_service.keep_original")) == bool
-
-
-def test_config_getfilepath(use_backup):
-    from src.helpers.config import config
-    import os.path
-    assert config.get_filepath("sync_service.original_dir") == os.path.expanduser("~/PySyncOCR/original")

